support/BRFileService.c

```diff
@@ -260,213 +599,181 @@ fileServiceLoad (BRFileService fs,
+        uint8_t  *entityBytes;
 
+        BRFileServiceHeaderFormatVersion headerVersion = dataBytes[offset];
+        offset += 1;
 
-    // Always, always write the header for the currentHeaderFormatVersion
+        switch (headerVersion) {
+            case HEADER_FORMAT_1:
+                version = dataBytes[offset];
+                offset += 1;
 
-    if (// write the header version
-        1 != fwrite(&currentHeaderFormatVersion, sizeof(BRFileServiceHeaderFormatVersion), 1, file) ||
-        // then the version
-        1 != fwrite (&entityType->currentVersion, sizeof(BRFileServiceVersion), 1, file) ||
-        // then the checksum?
-        // write the bytesCount
-        1 != fwrite(&bytesCount, sizeof (uint32_t), 1, file)) {
-        fileServiceFailedUnix (fs, bytes, file, errno);
-        return;
-    }
+                entityBytesCount = UInt32GetBE (&dataBytes[offset]);
+                offset += sizeof (uint32_t);
 
-    // write the bytes.
-    if (bytesCount != fwrite(bytes, 1, bytesCount, file)) {
-        fileServiceFailedUnix (fs, bytes, file, errno);
-        return;
+                break;
+        }
+
+        // Assert entityBytesCount remain in dataBytes
+        if (offset + entityBytesCount > dataBytesCount) {
+            assert (0); // In DEBUG builds.
+            return fileServiceFailedImpl (fs, 1, (dataBytes == dataBytesBuffer ? NULL : dataBytes), NULL,
+                                          "missed bytes count");
+        }
+
+        entityBytes = &dataBytes[offset];
+
+        switch (headerVersion) {
+            case HEADER_FORMAT_1:
+                // compute then compare checksum
+                break;
+        }
+
+        // Look up the entity handler
+        BRFileServiceEntityHandler *handler = fileServiceEntityTypeLookupHandler(entityType, version);
+        if (NULL == handler)
+            return fileServiceFailedImpl (fs, 1, (dataBytes == dataBytesBuffer ? NULL : dataBytes), NULL,
+                                          "missed type handler");
+
+        // Read the entity from buffer and add to results.
+        void *entity = handler->reader (handler->context, fs, entityBytes, entityBytesCount);
+        if (NULL == entity)
+            return fileServiceFailedEntity (fs, 1, (dataBytes == dataBytesBuffer ? NULL : dataBytes), NULL,
+                                            type, "reader");
+
+        // Update restuls with the newly restored entity
+        BRSetAdd (results, entity);
+
+        // If the read version is not the current version, update
+        if (updateVersion &&
+            (version != entityType->currentVersion ||
+             headerVersion != currentHeaderFormatVersion))
+            _fileServiceSave (fs, type, entity, 0);
     }
+    sqlite3_reset (fs->sdbSelectAllStmt);
+    pthread_mutex_unlock (&fs->lock);
 
-    if (0 != fclose (file)) {  fileServiceFailedUnix (fs, bytes, NULL, errno); return; }
+    if (dataBytes != dataBytesBuffer) free (dataBytes);
```

Comment:
For the early return cases in the while loop, is this deallocation required there too? Same with the sqlite3_reset?

Reply:
Early returns have the bytes freed if needed with:
```
            return fileServiceFailedEntity (fs, 1, (dataBytes == dataBytesBuffer ? NULL : dataBytes), NULL,
                                            type, "reader");
```

Reply:
Nice!